### PR TITLE
[T13-2]ForumBook

### PR DESCRIPTION
**Pull request by xllx1**

### CODE REVIEW

**Comment by ewaldhew** [comment]
**Comments for v1.2**

AboutUs: 
- Naming convention for team photos not followed. Use your Github username in _all lowercase_ and save as `.png`.

README:
- Remove CS2113-specific info such as the link to Learning Outcomes

DeveloperGuide:
- Good that the existing examples and diagrams are updated to your product name
- 3.1: Should be CRUD not CURD, minor issue though 😅 
- 3.3.1: Design consideration briefly mentioned but no evaluation of alternative solutions like the existing ones.
- 3.3.2: Some diagram could be useful here. How do these parts connect? What relationships have you defined between the different data?
- 3.3.3: Also a possible place to put diagram/example. Decide where best to put to enhance your explanations.
- 3.4: Need to add more details. Watch out for typos e.g. `UnitOfWork` becomes `UnitOfWord`.
- Reminder: Each member must have at least one diagram/example in their description of implementation details for their feature.
Maybe add class diagrams for specific new classes that are used in a feature, sequence diagrams to illustrate complicated flow of events, etc...

File: src/main/java/com/t13g2/forum/storage/forum/ForumBookStorage.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
+package com.t13g2.forum.storage.forum;
+
+/**
+ *
+ */
```
Comment (ewaldhew) [inline src/main/java/com/t13g2/forum/storage/forum/ForumBookStorage.java]: A lot of these empty header comments (add just to satisfy checkstyle right 😆), don't forget to fill them in later on!

File: src/main/java/com/t13g2/forum/storage/forum/IEvent.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.t13g2.forum.storage.forum;
+
+/**
+ *
+ */
+public interface IEvent {
+}
```
Comment (ewaldhew) [inline src/main/java/com/t13g2/forum/storage/forum/IEvent.java]: What might this be for? If it's to specify an event interface you could just extend from Java's existing `Event`

File: src/main/java/com/t13g2/forum/model/ForumBook.java [hunk 3]
```diff
@@ -27,12 +28,12 @@
         persons = new UniquePersonList();
     }
 
-    public AddressBook() {}
+    public ForumBook() {}
```
Comment (ewaldhew) [inline src/main/java/com/t13g2/forum/model/ForumBook.java]: Actually after looking at the code, you guys completely skipped this layer of abstraction right? Data from commands is written directly to storage via the `UnitOfWork` interface.
`ModelManager` etc. are not being used anymore in your project right now AFAICS, but that shouldn't be the case. Have a look at how `Model` component wraps around the storage.

File: src/main/java/com/t13g2/forum/storage/forum/ForumBookStorage.java (hunk 1 above)
Comment (Meowzz95) [reply src/main/java/com/t13g2/forum/storage/forum/ForumBookStorage.java]: Sure!

File: src/main/java/com/t13g2/forum/storage/forum/IEvent.java (hunk 2 above)
Comment (Meowzz95) [reply src/main/java/com/t13g2/forum/storage/forum/IEvent.java]: Okay will take a look at `Event`

File: src/main/java/com/t13g2/forum/model/ForumBook.java (hunk 3 above)
Comment (Meowzz95) [reply src/main/java/com/t13g2/forum/model/ForumBook.java]: Hi, for this, May I discuss with you about this? I've taken a brief look at `ModelManager`, this class defines what we can do with the data.
Eg:

* We can add a `Person` to storage
* We can update a `Person` to storage
* We can check if a `Person` exists 
* We CANNOT directly check if a `Person` with a specific attribute exists

This is what I mean by defining the behaviour.

This is a good design in AddressBook case since there is only one entity to manage, `Person`.

However, in our case, we have more entities, `Thread`, `Comment`, `User`. To better manage the data, `UnitOfWork` is implemented to manage them, it does not directly manage the data, instead, there is another layer of abstraction `EntityRepository`, these repositories interface defines how we can interact with data.

Eg:

* We can add `Announcement`
* We can add `User`
* We can get latest `Announcement`
* We CANNOT directly get `User` by its password

This is preciously what model manager does. Yes, we are not currently using `ModelManager`, because I don't see it fit. If I modify `ModelManager` to wrap `UnitOfWork`, it will end up more than 20 methods managing `Thread`, `Comment`, `User` etc. entities. This is obviously not a good design (single responsibility?).

TLDR:
`EntityRepository` is somehow similar to `ModelManager`, `UnitOfWork` is an additional layer of abstraction which takes care of data consistency(commit, rollback) and manages `EntityRepository`.

I'm not sure if I got this correctly, do let me know what I can improve on this. Thanks a lot for your effort.

Comment (Meowzz95) [reply src/main/java/com/t13g2/forum/model/ForumBook.java]: Hi good morning! @ewaldhew
